hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/SstFilteringService.java

```diff
@@ -0,0 +1,205 @@
+              <String, SnapshotInfo>> iterator = snapshotInfoTable
+              .iterator()) {
+        iterator.seekToFirst();
+
+        long snapshotLimit = snapshotLimitPerTask;
+
+        while (iterator.hasNext() && snapshotLimit > 0) {
+          Table.KeyValue<String, SnapshotInfo> keyValue = iterator.next();
+          String snapShotTableKey = keyValue.getKey();
+          SnapshotInfo snapshotInfo = keyValue.getValue();
+
+          File omMetadataDir =
+              OMStorage.getOmDbDir(ozoneManager.getConfiguration());
+          String snapshotDir = omMetadataDir + OM_KEY_PREFIX + OM_SNAPSHOT_DIR;
+          Path filePath =
+              Paths.get(snapshotDir + OM_KEY_PREFIX + FILTERED_SNAPSHOTS);
+
+          boolean isSnapshotFiltered = false;
+
+          // If entry for the snapshotID is present in this file,
+          // it has already undergone filtering.
+          if (Files.exists(filePath)) {
+            List<String> processedSnapshotIds = Files.readAllLines(filePath);
+            if (processedSnapshotIds.contains(snapshotInfo.getSnapshotID())) {
+              isSnapshotFiltered = true;
+            }
+          }
+
+          if (!isSnapshotFiltered) {
+            LOG.debug("Processing snapshot {} to filter relevant SST Files",
+                snapShotTableKey);
+
+            List<Pair<String, String>> prefixPairs =
+                constructPrefixPairs(snapshotInfo);
+
+            String dbName = OM_DB_NAME + snapshotInfo.getCheckpointDirName();
+
+            RDBStore rdbStore = (RDBStore) OmMetadataManagerImpl
+                .loadDB(ozoneManager.getConfiguration(), new File(snapshotDir),
+                    dbName, true);
+            RocksDatabase db = rdbStore.getDb();
+            db.deleteFilesNotMatchingPrefix(prefixPairs);
+
+            // mark the snapshot as filtered by writing to the file
+            String content = snapshotInfo.getSnapshotID() + "\n";
+            Files.write(filePath, content.getBytes(StandardCharsets.UTF_8),
+                StandardOpenOption.CREATE, StandardOpenOption.APPEND);
+            snapshotLimit--;
+            snapshotFilteredCount.getAndIncrement();
+          }
+        }
+      } catch (RocksDBException | IOException e) {
+        LOG.error("Error during Snapshot sst filtering ", e);
+      }
+
+      // nothing to return here
+      return BackgroundTaskResult.EmptyTaskResult.newResult();
+    }
+
+    /**
+     * @param snapshotInfo
+     * @return a list of pairs (tableName,keyPrefix).
+     * @throws IOException
+     */
+    private List<Pair<String, String>> constructPrefixPairs(
+        SnapshotInfo snapshotInfo) throws IOException {
+      String volumeName = snapshotInfo.getVolumeName();
+      String bucketName = snapshotInfo.getBucketName();
+
+      long volumeId = ozoneManager.getMetadataManager().getVolumeId(volumeName);
+      // TODO : buckets can be deleted via ofs , handle deletion of bucket case.
```

Comment:
I know it is not part of this PR, but I'm curious about bucket deletion.  Wouldn't that require the creation of new sst files?

Reply:
What I meant here is that since the background service doesn't acquire any bucket lock, and some one deletes it , it might fail to compute ` ozoneManager.getMetadataManager().getBucketId(volumeName, bucketName);` since it reads from active db bucket table currently in the code. Probably we should read it from the snapshot instead of active db .

Reply:
I totally misunderstood, thanks for clarifying